js-pkg/apps/demos/src/components/webrtc.js

```diff
@@ -0,0 +1,152 @@
+
+  React.useEffect(() => {
+    console.log('new connection')
+    let conn = new RTCPeerConnection({
+      iceServers: [{ urls: 'stun:stun.l.google.com:19302' }]
+    })
+
+    conn.onnegotiationneeded = async () => {
+      try {
+        console.log('sending offer')
+        makingOfferRef.current = true
+        await conn.setLocalDescription()
+        setSignalingMessages({ type: 'offer', sdp: conn.localDescription.sdp })
+      } catch (e) {
+        console.error(e)
+      } finally {
+        makingOfferRef.current = false
+      }
+    }
+
+    conn.onicecandidate = (e) => {
+      let candidate = e.candidate
+      setSignalingMessages({ type: 'candidate', candidate })
+    }
+
+    conn.onconnectionstatechange = () => {
+      console.log('current webrtc connection state: ', conn.connectionState)
+    }
+
+    conn.oniceconnectionstatechange = (_e) => {
+      if (conn.iceConnectionState === 'failed' || conn.iceConnectionState === 'disconnected') {
+        console.log('conn failed, restarting ICE')
+        conn.restartIce()
+      }
+    }
+
+    connRef.current = conn
+
+    for (let func of connSetupArray ?? []) {
+      func(conn)
+    }
+
+    return () => {
+      conn.close()
+    }
+  }, [])
+
+  React.useEffect(() => {
+    ;(async () => {
+      let conn = connRef.current
+      let makingOffer = makingOfferRef.current
+      const offerCollides = makingOffer || conn.signalingState !== 'stable'
+      const ignoreOffer = !isPeerPolite && offerCollides
+      let processed = 0
+      try {
+        for (let msg of signalingMessages) {
+          console.log(msg)
+          processed++
+          switch (msg.type) {
+            case 'answer':
+              await conn.setRemoteDescription(msg)
+              break
+            case 'offer':
+              if (ignoreOffer) break
+              await conn.setRemoteDescription(msg)
+              await conn.setLocalDescription()
+              setSignalingMessages({ type: 'answer', sdp: conn.localDescription.sdp })
+              break
+            case 'candidate':
+              try {
+                await conn.addIceCandidate(msg.candidate)
+              } catch (err) {
+                if (!ignoreOffer) throw err
+              }
+              break
+            default:
+              throw new Error('unknown message')
+          }
+        }
+      } catch (e) {
+        console.error(e)
+        console.error('failed conn manip')
+      } finally {
+        if (processed > 0) removeFromRecv(processed)
+      }
+    })()
+  }, [signalingMessages])
+
+  const getLatency = async () => {
+    const stats = await connRef.current?.getStats(null)
+
+    for (const [_, st] of stats ?? []) {
+      if (st.type === 'candidate-pair') {
+        if (typeof st.currentRoundTripTime === 'number') {
+          return st.currentRoundTripTime / 2
+        } else {
+          throw new Error('current round trip time not reported!')
+        }
+      }
+    }
+  }
```

Comment:
nit: right now this function is expected to either return a number, NaN, undefined, or throw an Error. Might be simpler to just have this return either a number or null?

I think that would just mean dropping the `throw new Error` block, changing the st.currentRoundTripTime check to `Number.isFinite(st.currentRoundTripTime)` and then adding a `return null` at the end.

Should simplify some of the calling code, too.

Reply:
ye done. thank you!

Reply:
Looks like this can still return `null` without a return at the end of the fn?

Reply:
ah that returns undefined not null
my bad